USF-OS/FogOS
# Merge ps implementation back to main

[Issue Link #51](https://github.com/USF-OS/FogOS/issues/51)

**Implemented:**

**1. kernel level changes**
- sys_getprocs -- new system call to support ps command
- getprocs(pd, max) -- new system call to support ps command where pd is a proc_data array and max is number of processes to report (see man page)

**2. user level changes**
- ps -- new user space command that implements a traditional unix process status of running processes
-  usertests -- added 4 test cases to xv6 regression test suite

**3. documentation changes**
- docs/ps.md -- man page in markdown format
- docs/getprocs.md -- man page in markdown format

**Implementation details**
- getprocs is implemented as system call 22 in the appropriate  kernel and user files
- the only args support in this implementation is "-h" which provides minimal help

**How to test**
Note: automated testing is incomplete due to lack of shell scripting support in this release of xv6

1. run "ps" in xv6: Note, repeated executions of 'ps' should result in the 3rd PID increasing:
```
$ ps
3 processes are running
PID	PPID	STATE	SIZE	NAME
1	0	sleep 	16384	init
2	1	sleep 	20480	sh
3	2	run   	16384	ps
$ ps
3 processes are running
PID	PPID	STATE	SIZE	NAME
1	0	sleep 	16384	init
2	1	sleep 	20480	sh
4	2	run   	16384	ps
```
2. run the regression suite: /usertests. Note this process runs somewhat slowly.
```
$ /usertests getprocstest
usertests starting
test getprocstest: getprocstest: getprocs test1 returned 4 procs
getprocstest: getprocs test2 returned 1 procs
getprocstest: getprocs test3 returned 4 procs
getprocstest: getprocs test OK
OK
ALL TESTS PASSED
``` 
- test1: run getprocs with default params: expect 4 procs
- test2: run getprocs with max=1: expect 1 proc 
- test3: run getprocs with max=NPROC+1 (exceed max value): expect 4 procs
- test4: run getprocs with invalid max, expect error
